package/kedro_viz/integrations/deployment/base_deployer.py

```diff
@@ -0,0 +1,108 @@
+"""`kedro_viz.integrations.deployment.base_deployer` defines
+creation of Kedro-viz build"""
+
+import abc
+import json
+import logging
+import tempfile
+from datetime import datetime
+from pathlib import Path
+
+from jinja2 import Environment, FileSystemLoader
+from packaging.version import parse
+
+from kedro_viz import __version__
+from kedro_viz.api.rest.responses import save_api_responses_to_fs
+from kedro_viz.integrations.kedro import telemetry as kedro_telemetry
+
+_HTML_DIR = Path(__file__).parent.parent.parent.absolute() / "html"
+_METADATA_PATH = "api/deploy-viz-metadata"
+
+logger = logging.getLogger(__name__)
+
+
+class BaseDeployer(abc.ABC):
+    """A class to handle the creation of Kedro-viz build.
+
+    Attributes:
+        _path (str): build path name.
+        _fs (fsspec.filesystem): Filesystem for local/remote protocol.
+
+    Methods:
+        deploy_and_get_url(): Deploy Kedro-viz to cloud storage provider/local and return its URL.
+    """
+
+    def __init__(self):
+        self._path = None
+        self._fs = None
+
+    def _upload_api_responses(self):
+        """Write API responses to the build."""
+        save_api_responses_to_fs(self._path)
+
+    def _ingest_heap_analytics(self):
+        """Ingest heap analytics to index file in the build."""
+        project_path = Path.cwd().absolute()
+        heap_app_id = kedro_telemetry.get_heap_app_id(project_path)
+        heap_user_identity = kedro_telemetry.get_heap_identity()
+        should_add_telemetry = bool(heap_app_id) and bool(heap_user_identity)
+        html_content = (_HTML_DIR / "index.html").read_text(encoding="utf-8")
+        injected_head_content = []
+
+        env = Environment(loader=FileSystemLoader(_HTML_DIR))
+
+        if should_add_telemetry:
+            logger.debug("Ingesting heap analytics.")
+            telemetry_content = env.get_template("telemetry.html").render(
+                heap_app_id=heap_app_id, heap_user_identity=heap_user_identity
+            )
+            injected_head_content.append(telemetry_content)
+
+        injected_head_content.append("</head>")
+        html_content = html_content.replace("</head>", "\n".join(injected_head_content))
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            temp_file_path = f"{temp_dir}/index.html"
+
+            with open(temp_file_path, "w", encoding="utf-8") as temp_index_file:
+                temp_index_file.write(html_content)
+
+            self._fs.put(temp_file_path, f"{self._path}/")
```

Comment:
Minor Nit: could we not do this directly in the path instead of copying from temp for efficiency?

Reply:
I can recollect, we had to do it this way because we had some issues getting this to work with s3fs